### ISSUE TITLE
SEVERE: initializeUniformWeight: failed when reading for file: /W5D_1000.dat [org.uma.jmetal.util.JMetalException <init>] java.lang.NullPointerException

### ISSUE THREAD

**Issue by ameera93** [jMetal/jMetal#237]
Facing this issue while running moead for 5 objectives wit 1000 population size. why its happening?

**Comment by matthieu-vergne** [comment]
Well, titles are not the best place to put traces. Could you provide the complete trace here, please?

**Comment by ameera93** [comment]
SEVERE: initializeUniformWeight: failed when reading for file: /W5D_20.dat [org.uma.jmetal.util.JMetalException <init>]
java.lang.NullPointerException

this is the one, although i have included all of the required source files to my project, it seems that its unable to locate the W5D_1000.dat file

**Comment by matthieu-vergne** [comment]
Is it really the whole stack trace? Seems rather short to me.

**Comment by ameera93** [comment]
2017-11-23 00:39:02.680 INFO: Loggers configured with null [org.uma.jmetal.util.JMetalLogger configureLoggers]
2017-11-23 00:39:02.726 INFO: ExecuteAlgorithms: Preparing output directory [org.uma.jmetal.util.experiment.component.ExecuteAlgorithms run]
2017-11-23 00:39:02.926 INFO:  Running algorithm: SPEA2, problem: 0, run: 0, funFile: C:/Users/Ameera/workspace/IOT_JMetal/src/20/Experiments/data/SPEA2/0/FUN0.tsv [org.uma.jmetal.util.experiment.util.ExperimentAlgorithm runAlgorithm]
2017-11-23 00:39:02.930 INFO:  Running algorithm: NSGAII, problem: 0, run: 0, funFile: C:/Users/Ameera/workspace/IOT_JMetal/src/20/Experiments/data/NSGAII/0/FUN0.tsv [org.uma.jmetal.util.experiment.util.ExperimentAlgorithm runAlgorithm]
2017-11-23 00:39:02.940 INFO:  Running algorithm: MOEAD, problem: 0, run: 0, funFile: C:/Users/Ameera/workspace/IOT_JMetal/src/20/Experiments/data/MOEAD/0/FUN0.tsv [org.uma.jmetal.util.experiment.util.ExperimentAlgorithm runAlgorithm]
2017-11-23 00:39:02.943 SEVERE: initializeUniformWeight: failed when reading for file: /W5D_20.dat [org.uma.jmetal.util.JMetalException <init>]
java.lang.NullPointerException
	at java.io.Reader.<init>(Unknown Source)
	at java.io.InputStreamReader.<init>(Unknown Source)
	at org.uma.jmetal.algorithm.multiobjective.moead.AbstractMOEAD.initializeUniformWeight(AbstractMOEAD.java:109)
	at org.uma.jmetal.algorithm.multiobjective.moead.MOEAD.run(MOEAD.java:46)
	at org.uma.jmetal.util.experiment.util.ExperimentAlgorithm.runAlgorithm(ExperimentAlgorithm.java:67)
	at org.uma.jmetal.util.experiment.component.ExecuteAlgorithms.lambda$run$0(ExecuteAlgorithms.java:42)
	at org.uma.jmetal.util.experiment.component.ExecuteAlgorithms$$Lambda$5/18968908.accept(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.ForEachOps$ForEachTask.compute(Unknown Source)
	at java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(Unknown Source)
	at java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)

Exception in thread "main" org.uma.jmetal.util.JMetalException
	at org.uma.jmetal.algorithm.multiobjective.moead.AbstractMOEAD.initializeUniformWeight(AbstractMOEAD.java:128)
	at org.uma.jmetal.algorithm.multiobjective.moead.MOEAD.run(MOEAD.java:46)
	at org.uma.jmetal.util.experiment.util.ExperimentAlgorithm.runAlgorithm(ExperimentAlgorithm.java:67)
	at org.uma.jmetal.util.experiment.component.ExecuteAlgorithms.lambda$run$0(ExecuteAlgorithms.java:42)
	at org.uma.jmetal.util.experiment.component.ExecuteAlgorithms$$Lambda$5/18968908.accept(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.ForEachOps$ForEachTask.compute(Unknown Source)
	at java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(Unknown Source)
	at java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)

**Comment by matthieu-vergne** [comment]
That's better. Just the line from "SEVERE" to the end is OK. The stuff after that line is the stack trace.

It fails [here](https://github.com/jMetal/jMetal/blob/9521559c11dbec3e348f56ae823b47c0baeba811/jmetal-algorithm/src/main/java/org/uma/jmetal/algorithm/multiobjective/moead/AbstractMOEAD.java#L109) because it cannot read the file, which is named based on [that](https://github.com/jMetal/jMetal/blob/9521559c11dbec3e348f56ae823b47c0baeba811/jmetal-algorithm/src/main/java/org/uma/jmetal/algorithm/multiobjective/moead/AbstractMOEAD.java#L104). What it means is that, based on the parameters you give, it will look for a file which contains the data, such that the name of the file respects this pattern:
```java
"W" + problem.getNumberOfObjectives() + "D_" + populationSize + ".dat"
```
Basically, you should first have the data file, then name it following this pattern, and only then configure the algorithm with the corresponding parameters before to run, as you did. If you have the data, just rename the file as requested. If you don't have the file, you will need to make it first.

By the way, it initialises by itself if you run for 2 objectives and with a population of less than 300 individuals. If you don't have any file to provide, then with this implementation you must restrict to these sizes. Don't ask me why, though.

So now, the big questions:
- do you have data?
- if not, can you restrict to 2 objectives and 300 individuals?

**Comment by ajnebro** [comment]
If you configure MOEA/D to solve a problem with 5 objectives and a population size of 1000, the code will look for a file name `W5D_1000.dat`, which is already located in the `jmetal-core/src/main/resources/MOEA_Weights` directory. I've just configured the `MOEADRunner` class to solve problem DTLZ1 with 5 objectives and a population = 1000 and it works.

However, you get an `initializeUniformWeight: failed when reading for file: /W5D_20.dat` error, with indicates that the file named W5D_20.dat cannot be found, which is true. Please, verify the population size that you are setting.

**Comment by ameera93** [comment]
sir, i tried for 5 objectives and 1000 population size. still its unable to locate W5D_1000.dat file. i can't figure  out the problem.

**Comment by ajnebro** [comment]
Are you using the `MOEADRunner` class to configure and run MOEA/D?

**Comment by ameera93** [comment]
hi antonio hope to find you in the best of health. I am here again to ask a
question that is why for MOEA/D algorithm with 5 objectives the weight
vector size starts from 1000 to onwards. Why there is no file for smaller
population size? and if i want to check the results for smaller population
size how can i introduce the weight vectors for smaller one. and in regard
of your previous email i am using experiment class for building and
configuring the algorithms. Is that a good approach or should I use the
runner class?


waiting for your kind reply

Regards
Ameera Naeem

**Comment by ajnebro** [comment]
Hi Ameera.
  The reason is that MOEA/D seems to work better with large population sizes, so for dealing with problems with 5 five objectives the number of vectors is 1000. The choices if you want to get a lower number of solutions is to generate the weight vector file with the proper combination of vectors and dimensions or to use the provided file and at the end take the number of solutions you need. This last approach was used in the MOEA/D-DRA flavor that won the CEC 2009 competition.

  If you take a look to the implementation of MOEA/D in the current working version of jMetal in GitHub (https://github.com/jMetal/jMetal/blob/master/jmetal-exec/src/main/java/org/uma/jmetal/runner/multiobjective/MOEADRunner.java), you can indicate the population size (300 in the example) and the result population size (100 in the example).

**Comment by ameera93** [comment]
if we create our own weight vector of size let say 210 i.e. W5D_210.dat
then where to place this file. in which folder? i am not getting this point.

**Comment by ajnebro** [comment]
You can indicate the folder as shown in the MOEADRunner.java program: by using the `setDataDirectory()` method of the `MOEADBuilder` class. We use by default the resources directory located in `jmetal-core/src/main/resources`. Thus, the "MOEAD_Weights" directory included in line 70 in class MOEADRunner it assumed to be in that directory.

**Comment by ameera93** [comment]
Hi Antonio
I have a question. Is NSGA-III fully implemented in JMetal version 5?

**Comment by ajnebro** [comment]
The current implementation is based on the code (version 1.1, if I remember well) developed by Tsung-Che Chiang: http://web.ntnu.edu.tw/~tcchiang/publications/nsga3cpp/nsga3cpp.htm

We not included the modifications of version 1.2.

**Comment by matthieu-vergne** [comment]
Please close this issue if the original problem with MOEAD is over.
Other topics should lead to open other issues to not mix everything.
